File: java-solutions/linkedList/AddTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
    }
}

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dumHead = new ListNode(0);
        ListNode curr = dumHead;
        int carry = 0;
        while(l1 != null || l2 != null) {
            int x = (l1 == null)? 0 : l1.val;
            int y = (l2 == null)? 0 : l2.val;
            int sum = x + y + carry;
            carry = sum / 10;
            curr.val = sum % 10;
            curr = curr.next;
            if( l1 != null) l1 = l1.next;
            if( l2 != null) l2 = l2.next;
        }
        if(carry > 0) {
            curr.next = new ListNode(carry);
        }
        return dumHead;
    }
}

class AddTwoNumbers {
    public static void main(String[] args) {
        ListNode l1 = new ListNode(2);
        l1.next = new ListNode(4);
        l1.next.next = new ListNode(3);
        ListNode l2 = new ListNode(5);
        l2.next = new ListNode(6);
        l2.next.next = new ListNode(4);

        Solution test = new Solution();
        test.addTwoNumbers(l1, l2);
        ListNode t = l1;
        t.val = 100;
        System.out.println( t.val);

    }
}

File: leetCode/linkedList/addTwoNumbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
const reverse = head => {
    let cur = head;
    let next = head.next;
    let prev = null;
    while (next !== null) {
        cur.next = prev;
        prev = cur;
        cur = next;
        next = next.next;
    }
    cur.next = prev;
    return cur;
}

var addTwoNumbers = function(l1, l2) {
    l1 = reverse(l1);
    l2 = reverse(l2);
    let carry = 0;
    const dummyHead = new ListNode(0);
    let pointer = dummyHead;
    while (l1 !== null || l2 !== null) {
        let sum = 0;
        if (l1 !== null) {
            sum += l1.val;
            l1 = l1.next;
        }
        if (l2 !== null) {
            sum += l2.val;
            l2 = l2.next;
        }
        if (carry > 0) {
            sum += carry;
        }
        carry = parseInt(sum / 10);
        sum %= 10;
        pointer.next = new ListNode(sum);
        pointer = pointer.next;
    }
    if (carry > 0) {
        pointer.next = new ListNode(carry);
        pointer = pointer.next;
    }
    return reverse(dummyHead.next);
};